test_util/cluster.py

```diff
@@ -184,6 +186,90 @@ def check_ssh(self, hosts):
         """
         self.ssher.remote_cmd(hosts, ['echo'])
 
+    def mesos_metrics_snapshot(self, host):
+        """Return a snapshot of the Mesos metrics for host."""
+        if host in self.masters:
+            port = 5050
+        else:
+            port = 5051
+
+        with self.ssher.tunnel(host) as tunnel:
+            return json.loads(
+                tunnel.remote_cmd([
+                    'curl', '--silent', '{}:{}/metrics/snapshot'.format(host.private_ip, port)
+                ]).decode('utf-8')
+            )
+
+    def zk_mode(self, host):
+        """Return the mode of the ZooKeeper instance on host."""
+        with self.ssher.tunnel(host) as tunnel:
+            stat_out = tunnel.remote_cmd([
+                'sudo', 'docker', 'run',
+                '--net=host',
+                '--entrypoint=/bin/sh',
+                'gophernet/netcat',
+                '-c', '"echo stat | nc localhost 2181"',
+            ])
+        for line in stat_out.split(b'\n'):
+            line = line.strip()
+            if line.startswith(b'Mode: '):
+                mode = line.split(b':')[1].strip()
+                if mode not in (b'leader', b'follower', b'standalone'):
+                    raise Exception('Unexpected ZooKeeper mode {} on host {}'.format(mode, host))
+                return mode.decode('utf-8')
+        else:
+            raise Exception('ZooKeeper mode not found on host {}'.format(host))
+
+    def master_zk_modes(self):
+        """Return (master, zk_mode(master)) for each master."""
+        master_modes = [(master, self.zk_mode(master)) for master in self.masters]
+
+        # Validate modes.
+        modes = [m[1] for m in master_modes]
+        if len(modes) == 1 and modes != ['standalone']:
+            raise Exception(
+                'The mode for a standalone ZooKeeper must be standalone. modes: {}'.format(repr(master_modes))
+            )
+        elif len(master_modes) > 1:
+            if any(mode not in ('leader', 'follower') for mode in modes):
+                raise Exception(
+                    'All ZooKeepers in an ensemble must be a leader or follower. modes: {}'.format(repr(master_modes))
+                )
+            if modes.count('leader') != 1:
+                raise Exception(
+                    'There must be exactly one leader in a ZooKeeper ensemble. modes: {}'.format(repr(master_modes))
+                )
+
+        return master_modes
+
+
+def run_docker_container_daemon(tunnel, container_name, image, docker_run_args=None):
+    """Run a Docker container with the given name on the host at tunnel."""
+    docker_run_args = docker_run_args or []
+    tunnel.remote_cmd(
+        ['sudo', 'docker', 'run', '--name', container_name, '--detach=true'] + docker_run_args + [image]
+    )
+
+
+def run_bootstrap_zookeeper(tunnel):
```

Comment:
That second example is a little different since it's running the container locally, not over SSH. I was able to at least DRY the docker args: https://github.com/dcos/dcos/pull/648/commits/320d7289949935e33a9c68403b05e624d9794bbd